### PR TITLE
Air-grapped Workloads Definition

### PR DESCRIPTION
**Pull request by dbw7**
This is the first in a set of PRs for the implementation of air-gapped workloads. This first adds the Hauler definition, updates the example YAML file, and adds validation and tests for the Hauler definition.

### CODE REVIEW

File: pkg/image/definition.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ type Definition struct {
 	APIVersion      string          `yaml:"apiVersion"`
 	Image           Image           `yaml:"image"`
 	OperatingSystem OperatingSystem `yaml:"operatingSystem"`
+	Hauler          Hauler          `yaml:"hauler"`
```
Comment (atanasdinov) [inline pkg/image/definition.go:19]: Hauler is the implementation - not the API contract we want to expose to the user. This should be `workloads` or something similar where it's clear what goes into it.

File: pkg/image/definition.go [hunk 2]
```diff
@@ -48,6 +49,28 @@ type Suma struct {
...
 
+type Hauler struct {
+	ContainerImages []ContainerImage `yaml:"images"`
+	HelmCharts      []HelmChart      `yaml:"charts"`
+	Files           []File           `yaml:"files"`
+}
+
+type ContainerImage struct {
+	Name string `yaml:"name"`
+	Key  string `yaml:"key"`
```
Comment (atanasdinov) [inline pkg/image/definition.go:60]: What type of key do we specify here?

File: pkg/image/definition.go [hunk 3]
```diff
@@ -48,6 +49,28 @@ type Suma struct {
...
+
+type HelmChart struct {
+	Name    string `yaml:"name"`
+	RepoURL string `yaml:"repoURL"`
+	Version string `yaml:"version"`
+}
+
+type File struct {
+	Name string `yaml:"name"`
+	Path string `yaml:"path"`
```
Comment (atanasdinov) [inline pkg/image/definition.go:71]: Can this be a local path or is it always an HTTP URL?

File: pkg/image/validation.go [hunk 4]
```diff
@@ -159,5 +165,95 @@ func validateSuma(os *OperatingSystem) error {
...
+
+func validateHelmCharts(charts []HelmChart) error {
+	seenCharts := make(map[string]bool)
+
+	for _, chart := range charts {
+		if chart.Name == "" {
+			return fmt.Errorf("no chart name defined")
+		}
+		if chart.RepoURL == "" {
+			return fmt.Errorf("no chart repository URL defined")
```
Comment (atanasdinov) [inline pkg/image/validation.go:223]: Can we include the name of the chart in this wrapper?

File: pkg/image/validation.go [hunk 5]
```diff
@@ -159,5 +165,95 @@ func validateSuma(os *OperatingSystem) error {
...
+		}
+		if chart.Version == "" {
+			return fmt.Errorf("no chart version defined")
```
Comment (atanasdinov) [inline pkg/image/validation.go:226]: Can we include the name of the chart in this wrapper?

File: pkg/image/validation.go [hunk 6]
```diff
@@ -159,5 +165,95 @@ func validateSuma(os *OperatingSystem) error {
...
+
+func validateFiles(files []File) error {
+	seenFiles := make(map[string]bool)
+
+	for _, file := range files {
+		if file.Name == "" {
+			return fmt.Errorf("no file name defined")
+		}
+		if file.Path == "" {
+			return fmt.Errorf("no file path defined")
```
Comment (atanasdinov) [inline pkg/image/validation.go:249]: Can we include the name of the file in this wrapper?

File: pkg/image/validation_test.go [hunk 7]
```diff
@@ -533,3 +533,266 @@ func TestValidateOperatingSystemSumaMissingActivationKey(t *testing.T) {
...
+				{
+					Name: "hello-world:latest",
+					Key:  "",
+				},
+				{
+					Name: "rgcrprod.azurecr.us/longhornio/longhorn-ui:v1.5.1",
+					Key:  "carbide-key.pub",
+				},
+				{
+					Name: "fake-image:latest",
```
Comment (atanasdinov) [inline pkg/image/validation_test.go:590]: This is not necessary since it is identical to the first case.

File: pkg/image/validation_test.go [hunk 8]
```diff
@@ -533,3 +533,266 @@ func TestValidateOperatingSystemSumaMissingActivationKey(t *testing.T) {
...
+				},
+				{
+					Name: "fake-image2-with-dupe-key:latest",
```
Comment (atanasdinov) [inline pkg/image/validation_test.go:593]: This is not necessary since it is identical to the second case.

File: pkg/image/validation_test.go [hunk 9]
```diff
@@ -533,3 +533,266 @@ func TestValidateOperatingSystemSumaMissingActivationKey(t *testing.T) {
...
+		},
+		{
+			name: "No Image Name Defined",
+			images: []ContainerImage{
+				{
+					Name: "",
+					Key:  "",
+				},
+				{
+					Key: "",
```
Comment (atanasdinov) [inline pkg/image/validation_test.go:606]: Identical to the previous case.

File: pkg/image/validation_test.go [hunk 10]
```diff
@@ -533,3 +533,266 @@ func TestValidateOperatingSystemSumaMissingActivationKey(t *testing.T) {
...
+				{
+					Name: "hello-world:latest",
+					Key:  "carbide-key.pub",
+				},
+			},
+			expectedErr: "duplicate container image found: 'hello-world:latest'",
+		},
+	}
+
+	// Test
```
Comment (atanasdinov) [inline pkg/image/validation_test.go:627]: I don't understand the purpose of these comments. What does it say which we don't already see?

**Review by atanasdinov — CHANGES_REQUESTED** [review]
We must use an alternative name for the section in the definition file. The rest looks fine to me. Happy to approve after this is resolved.

File: pkg/image/definition.go (hunk 1 above)
Comment (rdoxenham) [reply pkg/image/definition.go:19]: Exactly this ^^

File: pkg/image/definition.go (hunk 3 above)
Comment (dbw7) [reply pkg/image/definition.go]: resolving as a result of the removal of hauler file handling

File: pkg/image/validation.go (hunk 6 above)
Comment (dbw7) [reply pkg/image/validation.go]: resolving as a result of the removal of hauler file handling

File: pkg/image/definition.go (hunk 2 above)
Comment (dbw7) [reply pkg/image/definition.go:59]: Putting what we discussed here for visibility:
https://rancherfederal.github.io/hauler-docs/docs/introduction/quickstart#using-the-command-line
This key is for supply chain artifacts as shown in the link above.
Will rename `Key` to `SupplyChainKey`

**Review by atanasdinov — APPROVED** [review]
I'm good with this landing after the `hauler` references are completely dropped. Good work!

**Review by rdoxenham — APPROVED** [review]
LGTM, thanks!

File: pkg/image/validation.go [hunk 11]
```diff
@@ -159,5 +165,69 @@ func validateSuma(os *OperatingSystem) error {
...
+	err = validateHelmCharts(definition.EmbeddedArtifactRegistry.HelmCharts)
+	if err != nil {
+		return fmt.Errorf("error validating helm charts: %w", err)
+	}
+
+	return nil
+}
+
+func checkIfEmbeddedArtifactRegistryDefined(definition *Definition) bool {
+	return len(definition.EmbeddedArtifactRegistry.HelmCharts) == 0 && len(definition.EmbeddedArtifactRegistry.ContainerImages) == 0
```
Comment (jdob) [inline pkg/image/validation.go:189]: I may totally be misreading this, but this equates to "If both of these are zero, return that yes, embedded artifact registry is defined". It feels like the logic is backwards from the name. Am I just undercaffeinated?

**Review by jdob — CHANGES_REQUESTED** [review]
I'll flip to approve if it turns out I'm completely wrong on the check method

File: pkg/image/validation.go (hunk 11 above)
Comment (dbw7) [reply pkg/image/validation.go]: nah you're right, your logic makes more sense

**Review by jdob — APPROVED** [review]

**Pull request merged**